File: versions/v1.3.0_20260223_124525/market_forge_app/lib/services/ai_service.dart
import 'package:market_forge_app/services/api_service.dart';
import 'package:market_forge_app/config/app_config.dart';

/// AI service for interacting with EmpireBox AI agents
class AiService {
  final ApiService _apiService;

  AiService({ApiService? apiService})
      : _apiService = apiService ?? ApiService();

  /// Generate title suggestions based on photos and category
  Future<List<String>> generateTitleSuggestions({
    required List<String> photoUrls,
    String? category,
  }) async {
    if (!AppConfig.enableAiTitleSuggestions) {
      return [];
    }

    try {
      final response = await _apiService.post(
        '${AppConfig.aiServiceEndpoint}/generate-title',
        {
          'photoUrls': photoUrls,
          'category': category,
        },
      );

      return (response['suggestions'] as List<dynamic>)
          .map((e) => e as String)
          .toList();
    } catch (e) {
      print('Error generating title suggestions: $e');
      return [];
    }
  }

  /// Enhance product description using AI
  Future<String?> enhanceDescription({
    required String originalDescription,
    required String title,
    required String category,
  }) async {
    if (!AppConfig.enableAiDescriptionEnhancement) {
      return null;
    }

    try {
      final response = await _apiService.post(
        '${AppConfig.aiServiceEndpoint}/enhance-description',
        {
          'description': originalDescription,
          'title': title,
          'category': category,
        },
      );

      return response['enhancedDescription'] as String?;
    } catch (e) {
      print('Error enhancing description: $e');
      return null;
    }
  }

  /// Detect category from photos and title
  Future<String?> detectCategory({
    required List<String> photoUrls,
    String? title,
  }) async {
    if (!AppConfig.enableAiCategoryDetection) {
      return null;
    }

    try {
      final response = await _apiService.post(
        '${AppConfig.aiServiceEndpoint}/detect-category',
        {
          'photoUrls': photoUrls,
          'title': title,
        },
      );

      return response['category'] as String?;
    } catch (e) {
      print('Error detecting category: $e');
      return null;
    }
  }

  /// Suggest price based on similar listings
  Future<double?> suggestPrice({
    required String title,
    required String category,
    required String condition,
  }) async {
    if (!AppConfig.enableAiPriceSuggestions) {
      return null;
    }

    try {
      final response = await _apiService.post(
        '${AppConfig.aiServiceEndpoint}/suggest-price',
        {
          'title': title,
          'category': category,
          'condition': condition,
        },
      );

      return (response['suggestedPrice'] as num?)?.toDouble();
    } catch (e) {
      print('Error suggesting price: $e');
      return null;
    }
  }

  /// Analyze photos for quality and suggestions
  Future<Map<String, dynamic>> analyzePhotos({
    required List<String> photoUrls,
  }) async {
    try {
      final response = await _apiService.post(
        '${AppConfig.aiServiceEndpoint}/analyze-photos',
        {
          'photoUrls': photoUrls,
        },
      );

      return response;
    } catch (e) {
      print('Error analyzing photos: $e');
      return {
        'quality': 'unknown',
        'suggestions': [],
      };
    }
  }
}

File: versions/v1.2.0_20260223_124200/backend/app/services/ai_service.py
"""
AI service for interfacing with EmpireBox agents.
"""
from typing import Optional, Dict, Any
from sqlalchemy.ext.asyncio import AsyncSession
from app.models import User
from app.services.economic_service import EconomicService
from app.config import settings


class AIService:
    """
    Service for AI-powered features using EmpireBox agents.
    
    This would integrate with the empire_box_agents module when available.
    For now, provides mock implementations with economic tracking.
    """
    
    def __init__(self, user: User, db: Optional[AsyncSession] = None):
        self.user = user
        self.db = db
        # TODO: Initialize TokenManager and RequestRouter from empire_box_agents
        # self.token_manager = TokenManager(user_id=str(user.id), tier=user.tier)
        # self.router = RequestRouter(token_manager=self.token_manager)
    
    async def generate_description(self, product_info: Dict[str, Any]) -> Dict[str, Any]:
        """
        Generate product description using AI.
        
        Args:
            product_info: Dictionary with title, category, condition, photos
            
        Returns:
            Dictionary with description and tokens_used
        """
        # TODO: Implement actual AI generation
        # result = await self.router.route_request(
        #     task_type="content_generation",
        #     prompt=f"Write product description for: {product_info}",
        #     complexity="high"
        # )
        
        # Mock implementation
        description = f"High-quality {product_info.get('title', 'item')} in excellent condition."
        input_tokens = 50
        output_tokens = 100
        
        # Track economic cost if DB session available
        if self.db and settings.economic_enabled:
            economic_service = EconomicService(self.db)
            await economic_service.record_ai_cost(
                entity_type="user",
                entity_id=self.user.id,
                input_tokens=input_tokens,
                output_tokens=output_tokens,
                model="gpt-4",
                description="Generate product description"
            )
        
        return {
            "description": description,
            "tokens_used": input_tokens + output_tokens,
            "input_tokens": input_tokens,
            "output_tokens": output_tokens
        }
    
    async def enhance_description(self, current_description: str) -> Dict[str, Any]:
        """
        Enhance existing product description.
        
        Args:
            current_description: Current description text
            
        Returns:
            Dictionary with enhanced_description and tokens_used
        """
        # Mock implementation
        enhanced = f"{current_description}\n\nEnhanced with additional details and SEO keywords."
        input_tokens = len(current_description.split()) * 2  # Rough estimate
        output_tokens = 60
        
        # Track economic cost if DB session available
        if self.db and settings.economic_enabled:
            economic_service = EconomicService(self.db)
            await economic_service.record_ai_cost(
                entity_type="user",
                entity_id=self.user.id,
                input_tokens=input_tokens,
                output_tokens=output_tokens,
                model="gpt-4",
                description="Enhance product description"
            )
        
        return {
            "enhanced_description": enhanced,
            "tokens_used": input_tokens + output_tokens,
            "input_tokens": input_tokens,
            "output_tokens": output_tokens
        }
    
    async def suggest_price(self, product_info: Dict[str, Any]) -> Dict[str, Any]:
        """
        Suggest pricing for a product.
        
        Args:
            product_info: Dictionary with title, category, condition, description
            
        Returns:
            Dictionary with suggested_price, price_range, reasoning
        """
        # Mock implementation
        input_tokens = 80
        output_tokens = 40
        
        # Track economic cost if DB session available
        if self.db and settings.economic_enabled:
            economic_service = EconomicService(self.db)
            await economic_service.record_ai_cost(
                entity_type="user",
                entity_id=self.user.id,
                input_tokens=input_tokens,
                output_tokens=output_tokens,
                model="gpt-4",
                description="Suggest product price"
            )
        
        return {
            "suggested_price": 49.99,
            "price_range": {"min": 39.99, "max": 59.99},
            "reasoning": "Based on similar items and market conditions",
            "tokens_used": input_tokens + output_tokens,
            "input_tokens": input_tokens,
            "output_tokens": output_tokens
        }
    
    async def draft_message_response(
        self,
        message_content: str,
        context: Optional[str] = None
    ) -> str:
        """
        Generate AI draft response for a message.
        
        Args:
            message_content: Original message content
            context: Optional additional context
            
        Returns:
            Draft response text
        """
        # Mock implementation
        input_tokens = len(message_content.split()) * 2
        output_tokens = 50
        
        # Track economic cost if DB session available
        if self.db and settings.economic_enabled:
            economic_service = EconomicService(self.db)
            await economic_service.record_ai_cost(
                entity_type="user",
                entity_id=self.user.id,
                input_tokens=input_tokens,
                output_tokens=output_tokens,
                model="gpt-4",
                description="Draft message response"
            )
        
        return f"Thank you for your interest! {message_content[:50]}... I'll get back to you shortly."
